src/Streamer.cpp

```diff
@@ -43,102 +53,129 @@ FileWriter::Streamer::connect(std::string TopicName) {
+        return ProcessMessageResult::OK;
       }
-      RunStatus = IsConnected.get();
+      auto Temp = ConsumerCreated.get();
+      RunStatus = Temp.first;
+      Consumer = std::move(Temp.second);
+    } else {
+      throw std::runtime_error(
+          "Failed to set-up process for creating consumer.");
     }
+  } catch (std::runtime_error &Error) {
+    throw Error;
   } catch (std::exception &Error) {
-    LOG(Sev::Critical, "{}", Error.what());
+    LOG(Sev::Critical, "Got an exception when waiting for connection: {}",
+        Error.what());
+    throw Error;
   }
 
   // make sure that the connection is ok
   // attention: connect() handles exceptions
-  if (!RunStatus.connectionOK()) {
+  if (RunStatus < StreamerStatus::IS_CONNECTED) {
     throw std::runtime_error(Err2Str(RunStatus));
   }
 
-  // consume message and make sure that's ok
+  // Consume message and exit if we are beyond a message timeout
   KafkaW::PollStatus Poll = Consumer->poll();
   if (Poll.isEmpty() || Poll.isEOP()) {
-    if ((Options.StopTimestamp.count() > 0) &&
-        (systemTime() > (Options.StopTimestamp + Options.AfterStopTime))) {
-      LOG(Sev::Info, "Close topic {} after time expired",
-          MessageProcessor.topic());
+    if ((Options.StopTimestamp.count() > 0) and
+        (Options.ConsumerTimeout < (systemTime() - LastMessageTimestamp))) {
+      LOG(Sev::Info, "Stop stream timeout for topic \"{}\" reached. {} ms "
+                     "passed since last message received.",
+          MessageProcessor.topic(),
+          (systemTime() - LastMessageTimestamp).count());
       Sources.clear();
-      return ProcessMessageResult::STOP();
+      return ProcessMessageResult::STOP;
     }
-    return ProcessMessageResult::OK();
+    return ProcessMessageResult::OK;
   }
   if (Poll.isErr()) {
-    return ProcessMessageResult::ERR();
+    return ProcessMessageResult::ERR;
   }
 
-  // convert from KafkaW to Msg
-  Msg Message(Msg::fromKafkaW(Poll.isMsg()));
-  if (Message.type == MsgType::Invalid) {
-    return ProcessMessageResult::ERR();
+  // Convert from KafkaW to FlatbufferMessage, handles validation of flatbuffer
+  auto KafkaMessage = Poll.isMsg();
+  std::unique_ptr<FlatbufferMessage> Message;
+  try {
+    Message.reset(new FlatbufferMessage(
+        reinterpret_cast<const char *>(KafkaMessage->data()),
+        KafkaMessage->size()));
+  } catch (std::runtime_error &Error) {
+    LOG(Sev::Warning,
+        "Message that is not a valid flatbuffer encountered. The error was: {}",
+        Error.what());
+    return ProcessMessageResult::ERR;
   }
 
-  // Make sure that the message source is relevant and that the message is in
-  // the correct time window
-  DemuxTopic::DT MessageTime =
-      MessageProcessor.time_difference_from_message(Message);
-  if (std::find(Sources.begin(), Sources.end(), MessageTime.sourcename) ==
+  LastMessageTimestamp = systemTime(); // For dealing with messsage timeouts
```

Comment:
Add todo and JIRA issue about possible performance penalties of checking the system time.